modules/mod-deep-learning/ModelCard.h

```diff
@@ -0,0 +1,212 @@
+               .Format(m_msg);}
+
+   const TranslatableString m_msg;
+   const std::string m_trace;
+};
+
+namespace parsers 
+{
+   Doc ParseString(const std::string &json);
+   Doc ParseFile(const std::string &path);
+}
+
+// validators that accept a default value are no-throw
+namespace validators
+{
+
+   std::string tryGetString(const std::string &key, const Doc& doc);
+   std::string tryGetString(const std::string &key, const Doc& doc, const std::string &defaultValue);
+
+   int tryGetInt(const std::string &key, const Doc& doc);
+   int tryGetInt(const std::string &key, const Doc& doc, int defaultValue);
+
+   std::vector<std::string> tryGetStringArray(const std::string &key, const Doc& doc, bool lower=false);
+                                             
+   uint64_t tryGetUint64(const std::string &key, const Doc& doc);
+   uint64_t tryGetUint64(const std::string &key, const Doc& doc, uint64_t defaultValue);
+
+}
+
+class DeepModelManager;
+
+class ModelCard final
+{
+public: 
+   ModelCard() = default;
+   ModelCard(const ModelCard&) = default;
+
+   // returns {author}/{name}
+   std::string GetRepoID() const;
+
+   bool IsSame(const ModelCard& other) const { return (*this) == other; };
+
+private:
+   friend class DeepModelManager;
+   
+   // throws InvalidModelCardDocument if the given json is not valid. 
+   void DeserializeFromFile(const std::string& path, const Doc& schema);
+   void SerializeToFile(const std::string& path) const;
+
+   template < typename Writer >
+   void Serialize(Writer &writer) const;
+   void Deserialize(const Doc& doc, const Doc& schema);
+
+   void Validate(const Doc& doc, const Doc& schema);
+
+   bool IsLocal() { return m_is_local; }
+   void SetLocal(bool local) { m_is_local = local; }
+
+   std::string GetLocalPath() const { return m_local_path; }
+   void SetLocalPath(const std::string& path) { m_local_path = path; }
+
+public:
+
+   std::string name() const { return m_name; }
+   void name(const std::string &name) { m_name = name; }
+
+   std::string author() const { return m_author; }
+   void author(const std::string &author) { m_author = author; }
+
+   std::string long_description() const { return m_long_description; }
+   void long_description(const std::string &long_description) { m_long_description = long_description; }
+
+   std::string short_description() const { return m_short_description; }
+   void short_description(const std::string &short_description) { m_short_description = short_description; }
+
+   int sample_rate() const { return m_sample_rate; }
+   void sample_rate(int rate) { m_sample_rate = rate; }
+
+   bool multichannel() const { return m_multichannel; }
+   void multichannel(bool multichannel) { m_multichannel = multichannel; }
+
+   std::string effect_type() const { return m_effect_type; }
+   void effect_type(const std::string &type) { m_effect_type = type; }
+
+   const std::vector<std::string> domain_tags() const { return m_domain_tags; }
```

Comment:
This, and the following two vector-valued accessors, might return const reference, and the places that use them might bind variables with `const auto &tags = ...`, for avoidance of vector copies.

Reply:
358f92a21c5048518f61d8a9e5c83045659791ad